### ISSUE TITLE
Dimensional analysis

### ISSUE THREAD

**Issue by murisi** [anoma/namada#2247]
Namada currently deals with token amounts (which do not necessarily have integral values) as follows:
* Parse a user's decimal number into a `DenominatedAmount` and the token its quantifying
* Validate a denominated amount to ensure it does not have too many decimal places
* Increase the precision of the `DenominatedAmount` to match the denomination of the specified token
* In some instances (like ERC20), the denomination is discarded here leaving only an integral `Amount` in the transaction
* Sign the transaction either on the hardware wallet or in software and send to protocol
* If the protocol receives a `DenominatedAmount`, then it ignores the denomination and extracts the integral `Amount
  * If a function that receives a `DenominatedAmount` like `tx::transfer` or `tx::mint` is required, we sometimes attach an invalid (?) denomination in the knowledge that it will be immediately discarded by the function

**Comment by cwgoes** [comment]
Thanks for this write-up. I'm surprised that we use `DenominatedAmount`s in the protocol at all, other than perhaps in the MASP. (4) and (6) in particular sound strange and risky.

To me at least denomination is purely a display convention. Even for the MASP, different tokens should _not_ have different denominations; we should just use a maximum of four u64s to represent the full u256 range. We can keep a map of standardized denominations in storage - a given token identifier (including ERC20s) should have only one denomination display convention, set when the token is created. I think this more or less fits with your proposal - including `DenominatedAmounts` in transaction data is alright with me, as long as we always just immediately convert it to the underlying u256 Amount type.

**Comment by sug0** [comment]
My opinion is that these exploits are only client-exploits. The protocol only stores token amounts (i.e. 256-bit unsigned integer values) under user accounts. However, precisely because the protocol doesn't care about denoms, those should be removed from all protocol paths.

The flow should be something like:
1. Input formatted amount in some interface
2. Query the denomination of the token whose denominated amount was parsed
3. If the token is correctly denominated, extract the validated amount and send it over to the protcol in the form of a tx

This more or less matches the suggestion in your last bullet point. The hardware wallet could receive a denomination (maybe?) as an additional tx data field, and strip it out before signing. It's a hacky approach though.

**Comment by cwgoes** [comment]
@murisi @sug0 Do you know the current state here? Do we use `DenominatedAmount` anywhere in the protocol? We should conduct an internal audit for this.

**Comment by sug0** [comment]
yeah, `DenominatedAmount` values are still used in a couple of places

**Comment by cwgoes** [comment]
Do we have a list of exactly where it's used in the protocol? IMO in principle we don't want protocol logic to be denomination-dependent at all, denomination should be a display convention only.

**Comment by sug0** [comment]
> Do we have a list of exactly where it's used in the protocol? IMO in principle we don't want protocol logic to be denomination-dependent at all, denomination should be a display convention only.

@cwgoes not exactly, but you can kind of see where it's being used if you have [`ripgrep`](https://github.com/BurntSushi/ripgrep) installed, and run `rg DenominatedAmount crates` at the root of the repo
